sdks/python/apache_beam/runners/trivial_runner.py

```diff
@@ -0,0 +1,415 @@
+      for element in input_elements:
+        for window in element.windows:
+          key, value = element.value
+          grouped[window, key_coder.encode(key)].append(value)
+      output_elements = [
+          windowed_value.WindowedValue(
+              (key_coder.decode(encoded_key), values),
+              window.end, [window],
+              trigger.BatchGlobalTriggerDriver.ONLY_FIRING)
+          for ((window, encoded_key), values) in grouped.items()
+      ]
+    else:
+      # This handles generic merging and triggering.
+      trigger_driver = trigger.create_trigger_driver(
+          execution_state.windowing_strategy(input_pcoll), True)
+      grouped_by_key = collections.defaultdict(list)
+      for element in input_elements:
+        key, value = element.value
+        grouped_by_key[key_coder.encode(key)].append(element.with_value(value))
+      output_elements = []
+      for encoded_key, windowed_values in grouped_by_key.items():
+        for grouping in trigger_driver.process_entire_key(
+            key_coder.decode(encoded_key), windowed_values):
+          output_elements.append(grouping)
+
+    # Store the grouped values in the output PCollection.
+    output_coder = execution_state.windowed_coder(output_pcoll)
+    execution_state.set_pcollection_contents(
+        output_pcoll, [encode_all(output_elements, output_coder)])
+
+  def supported_requirements(self) -> Iterable[str]:
+    # Nothing non-trivial is supported.
+    return []
+
+
+class ExecutionState:
+  """A helper class holding various values and context during execution."""
+  def __init__(self, optimized_pipeline):
+    self.optimized_pipeline = optimized_pipeline
+    self._pcollections_to_encoded_elements = {}
+    self._counter = 0
+    self._process_bundle_descriptors = {}
+    # This emulates a connection to an SDK worker (e.g. its data, control,
+    # etc. channels).
+    # There are other variants available as well (e.g. GRPC, Docker, ...)
+    self.worker_handler = worker_handlers.EmbeddedWorkerHandler(
+        None,
+        state=worker_handlers.StateServicer(),
+        provision_info=None,
+        worker_manager=self)
+    self._windowed_coders = {}
+    # Populate all windowed coders before creating _pipeline_context.
+    for pcoll_id in self.optimized_pipeline.components.pcollections.keys():
+      self.windowed_coder_id(pcoll_id)
+    self._pipeline_context = pipeline_context.PipelineContext(
+        optimized_pipeline.components)
+
+  def register_process_bundle_descriptor(
+      self, process_bundle_descriptor: beam_fn_api_pb2.ProcessBundleDescriptor):
+    self._process_bundle_descriptors[
+        process_bundle_descriptor.id] = process_bundle_descriptor
+
+  def get_pcollection_contents(self, pcoll_id: str) -> List[bytes]:
+    return self._pcollections_to_encoded_elements[pcoll_id]
+
+  def set_pcollection_contents(self, pcoll_id: str, chunks: List[bytes]):
+    self._pcollections_to_encoded_elements[pcoll_id] = chunks
```

Comment:
`_pcollections_to_encoded_chunks` might be slightly more correct? I got tripped up because I thought each value in these lists encoded a single element.

Not sure the suggested name is really that much better though. Feel free to ignore

Reply:
I can see how this is non-obvious. I like your suggestion.